### ISSUE TITLE
Get deadlocked occasionally...

### ISSUE THREAD

**Issue by scott-xu** [IronLanguages/ironpython3#1596]
### Description

A worker thread is calling `PythonContext.CreateBuiltinModule` which locks `PythonContext`. Then it will call `FunctionCode.LazyCompileFirstTarget` where it tries to acquire `_CodeCreateAndUpdateDelegateLock`

PythonContext.cs
```C#
internal PythonModule GetBuiltinModule(string name) {
    lock (this) {
        PythonModule mod = CreateBuiltinModule(name);
        if (mod != null) {
            PublishModule(name, mod);
            return mod;
        }

        return null;
    }
}
```
FunctionCode.cs
```C#
internal void LazyCompileFirstTarget(PythonFunction function) {
    lock (_CodeCreateAndUpdateDelegateLock) {
        UpdateDelegate(function.Context.LanguageContext, true);
    }
}
```

A GC thread is executing the `WeakRef`'s finalizer which calls `FunctionCode.LazyCompileFirstTarget` and locks `_CodeCreateAndUpdateDelegateLock`, then it will call `PythonContext.DeleteIndex` where it tries to acquire `PythonContext`.

PythonContext.cs
```C#
internal PythonDeleteIndexBinder/*!*/ DeleteIndex(int argCount) {
    if (_deleteIndexBinders == null) {
        Interlocked.CompareExchange(ref _deleteIndexBinders, new PythonDeleteIndexBinder[argCount + 1], null);
    }

    lock (this) {
        if (_deleteIndexBinders.Length <= argCount) {
            Array.Resize(ref _deleteIndexBinders, argCount + 1);
        }

        if (_deleteIndexBinders[argCount] == null) {
            _deleteIndexBinders[argCount] = new PythonDeleteIndexBinder(this, argCount);
        }

        return _deleteIndexBinders[argCount];
    }
}
```

#### Worker Thread
![image](https://user-images.githubusercontent.com/1913505/202982117-5d399b7a-c229-42c5-8c7f-bde279c4d71f.png)
.
.
.
![image](https://user-images.githubusercontent.com/1913505/202982306-1e031722-d827-4576-a352-7c65f6234466.png)

#### GC Thread
![image](https://user-images.githubusercontent.com/1913505/202981821-909cac7c-794a-4343-b8ef-a5da0944830a.png)

**Comment by slozier** [comment]
Thanks for the report. I don't see a reason why we couldn't use a different lock in `PythonContext.DeleteIndex` (as well as in `GetIndex` and `SetIndex`). Do you have a code sample code that could be used to reproduce the issue so I can set up a unit test?

**Comment by scott-xu** [comment]
Here's the sample code.
```python
import xml.etree.ElementTree as ET
s = '<Data>' + ('1'*8192) + '</Data>'
elem = ET.fromstring(s)
result_len = len(ET.tostring(elem))
string_len = len(s)
result = str(result_len == string_len)
```

**Comment by slozier** [comment]
Thanks, was able to reproduce the deadlock with code like this:
```c#
var code = @"
import xml.etree.ElementTree as ET
s = '<Data>' + ('1'*8192) + '</Data>'
elem = ET.fromstring(s)
ET.tostring(elem)";

// retry to increase chance of deadlock
for (var i = 0; i < 100; i++) {
    Console.WriteLine(i);
    ScriptEngine eng = Python.CreateEngine();
    eng.SetSearchPaths(new List<string> { pathToLib });
    await Task.Run(() => eng.Execute(code));
}
```

**Comment by scott-xu** [comment]
Thanks for the quick fix!
Will the fix be included in 3.4 final release? It has been half a year since 3.4 beta1. When do you think 3.4 final would be ready?

**Comment by scott-xu** [comment]
Perfect! Thanks for the great work!
